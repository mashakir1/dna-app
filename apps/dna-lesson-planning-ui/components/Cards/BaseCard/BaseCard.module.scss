.card {
  background-color: $colour_white;
  border-radius: 16px;
}

.bordered {
  border: $colour_background_grey solid 1px;
}
