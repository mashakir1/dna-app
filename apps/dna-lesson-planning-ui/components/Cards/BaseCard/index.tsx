// Imports
import { ReactNode } from "react";

// Styles
import styles from "./BaseCard.module.scss";

interface BaseCardProps {
  bordered?: boolean;
  children: ReactNode;
  classNames?: string
  onClick?: () => void
}

export const BaseCard: (props: BaseCardProps) => JSX.Element = ({
  bordered, children, classNames, onClick
}) => (
  <div className={`${
    bordered ? styles.bordered : ""
  } ${
    styles.card
  } ${
    classNames ? classNames : ""
  }`} onClick={onClick}>{children}</div>
);
