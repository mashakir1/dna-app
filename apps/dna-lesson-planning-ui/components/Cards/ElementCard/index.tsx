// Imports
import { useCallback, useMemo, useState } from "react";

// Components
import { InfoIcon } from "components/Svg";
import { ClickableIcon } from "components/Icons";
import { IconSelector } from "./IconSelector";

// Styles
import styles from "./elementCard.module.scss";

/*
* How to use the ElementCard:
*  title, text, duration & subjectColour all show up as content on the card
*  selected determines if the card has been chosen by a user
*  onClick determines what happens when a user clicks on the card.
*    it also applies hover states
*  stickyCross will keep the crossIcon shown at all times
*  crossOnClick determines what will happen when the cross is clicked
*  infoOnClick determines what will happen when the infoIcon is clicked
*/

export interface ElementCardProps {
  id: string;
  title: string;
  text: string;
  duration: number;
  subjectColour: string;
  selected?: boolean;
  onClick?: (id: string) => void;
  stickyCross?: boolean;
  crossOnClick?: (id: string) => void;
  infoOnClick?: () => void;
}

export const ElementCard: (props: ElementCardProps) => JSX.Element = ({
  id, title, text, duration, subjectColour,
  selected, onClick, stickyCross, crossOnClick, infoOnClick
}) => {
  const clickable = useMemo(() => !!onClick, [onClick]);
  const [hovering, setHovering] = useState<boolean>(false);

  const onClickHandler = useCallback(() => onClick(id), [id, onClick]);
  const crossOnClickHandler = useCallback(() => crossOnClick(id), [id, crossOnClick]);

  return <div
    className={`${styles.elementCard} ${clickable ? styles.clickable : ""} flex column`}
    {...(onClick && { onClick: onClickHandler })}
    onMouseEnter={() => setHovering(true)}
    onMouseLeave={() => setHovering(false)}
  >
    <ElementCardHeader
      title={title}
      subjectColour={subjectColour}
      hovering={hovering}
      selected={selected}
      clickable={clickable}
      stickyCross={stickyCross}
      {...(crossOnClick && { crossOnClick: crossOnClickHandler })}
    />
    <ElementCardBody
      text={text}
      duration={duration}
      infoOnClick={infoOnClick}
    />
  </div>;
};

interface ElementCardHeaderProps {
  title: string;
  subjectColour: string;
  hovering: boolean;
  selected: boolean;
  clickable: boolean;
  stickyCross?: boolean;
  crossOnClick?: () => void;
}

const ElementCardHeader: (props: ElementCardHeaderProps) => JSX.Element = ({
  title, subjectColour, hovering,
  selected, clickable, stickyCross, crossOnClick
}) => {
  const headerColour = selected ? "#30CCD6" : subjectColour;

  return <div
    className={`${styles.elementCardHeader} flex justify-between items-center`}
    style={{ backgroundColor: headerColour }}
  >
    <p><b>{title}</b></p>
    <IconSelector
      hovering={hovering}
      selected={selected}
      clickable={clickable}
      stickyCross={stickyCross}
      crossOnClick={crossOnClick}
    />
  </div>;
};

interface ElementCardBodyProps {
  text: string;
  duration: number;
  infoOnClick: () => void;
}

const ElementCardBody: (props: ElementCardBodyProps) => JSX.Element = ({
  text, duration, infoOnClick
}) => {
  const handleInfoIconClick = (event) => {
    event.stopPropagation();
    infoOnClick();
  };

  return <div className={`${styles.elementCardBody} flex column justify-between`}>
    <p>{text}</p>
    <div className={`${styles.elementCardFooter} flex justify-between items-center`}>
      <p className={`${styles.duration} button small`}><b>{duration} mins</b></p>
      {infoOnClick && <ClickableIcon icon={<InfoIcon />} onClick={handleInfoIconClick} />}
    </div>
  </div>;
};
