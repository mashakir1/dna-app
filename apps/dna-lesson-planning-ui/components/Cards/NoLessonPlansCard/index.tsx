// Imports
import useTranslation from "next-translate/useTranslation";
import Link from "next/link";
import Trans from "next-translate/Trans";

// Components
import { BaseCard } from "../BaseCard";
import { IconWithBackground } from "components/Icons";
import { FileIcon } from "components/Svg";

// Styles
import styles from "styles/pages/home.module.scss";

export const NoLessonPlansCard = () => {
  const { t } = useTranslation("common");

  return <BaseCard classNames={`${styles.emptyCard} flex column items-center gap-m`}>
    <IconWithBackground icon={<FileIcon />} />
    <p>{t("no-lesson-plans-to-show")}</p>
    <p><Trans
      i18nKey={"need-help"}
      components={[<Link href={"/"} key={""} />]}
      ns={"common"}
    /></p>
  </BaseCard>;
};
