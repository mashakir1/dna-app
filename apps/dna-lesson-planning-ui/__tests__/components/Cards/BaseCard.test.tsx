import { render } from "@testing-library/react";

import { BaseCard } from "components/Cards/BaseCard";

describe("BaseCard", () => {
  it("should render an empty base card", () => {
    const { container } = render(
      <BaseCard>
        <></>
      </BaseCard>
    );
    expect(container).toMatchSnapshot();
  });
  it("should render a base card with text", () => {
    const { container } = render(
      <BaseCard>
        <p>BaseCard</p>
      </BaseCard>
    );
    expect(container).toMatchSnapshot();
  });
});
