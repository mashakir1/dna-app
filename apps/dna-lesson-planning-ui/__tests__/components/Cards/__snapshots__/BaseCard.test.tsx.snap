// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCard should render a base card with text 1`] = `
<div>
  <div
    class=" card "
  >
    <p>
      BaseCard
    </p>
  </div>
</div>
`;

exports[`BaseCard should render an empty base card 1`] = `
<div>
  <div
    class=" card "
  />
</div>
`;
